### ISSUE TITLE
rpm_packages table bdb/librpm corrupts RPM's DB on RHEL/CentOS 7

### ISSUE THREAD

**Issue by tomlanyon** [facebook/osquery#3178]
I believe this is the same as #2623, but that's marked closed and I wasn't sure if anyone would be watching comments there to re-open it.

tl;dr: on Enterprise Linux 7 (I'm testing on a 7.3 machine), running osqueryd or osqueryi as root seemingly modifies the RPM Berkeley DB's shared memory regions such that it's incompatible with the libdb and librpm shipped with the OS, breaking the yum and rpm utilities and anything else using the shared OS libraries.

AIUI the reason that #2623 saw different behavior between osqueryd and osqueryi is that osqueryd runs as root and (in this circumstance, I assume) osqueryi was running as a non-root user.  librpm, when opening the Berkeley DB, attempts to use shared memory regions backed by mmap'ing the /var/lib/rpm/__db.xxx files (which should be uid root, gid root and 0644 perms);  if running as non-root, opening the DB using these file-backed regions fails with permission errors and librpm retries after setting the DB_PRIVATE flag, which uses malloc'd private process memory for the database shared memory regions. More info on the BDB shared memory regions is in [the docs](https://docs.oracle.com/cd/E17276_01/html/programmer_reference/env_region.html).

When an osquery process runs as root and the rpm_packges table is queried, it _can_ open the filesystem backed shared memory files, and it modifies them / re-writes them with whatever version of libdb is linked into osquery at build time (db 6.1.26 at time of writing);  this then causes the OS's rpm tools (like the /bin/rpm binary, linked with libdb 5.3.21 on this EL7 machine) to fail to open the RPM DB.

For some reason this doesn't seem to be a problem on the Enterprise Linux 6 machines that I tested, which are running libdb 4.7.25. ¯\_(ツ)_/¯

Options I've thought of so far for how this could be resolved:
- use the OS's shared libdb rather than statically linking some other version of libdb; this will always ensure compatibility with the OS.
- statically link an older version of libdb, to be compatible with whatever the OS ships; might require different versions to be built/linked for different Enterprise Linux OS versions.
- drop privileges in the rpm_packages table implementation to ensure it can't write to the /var/lib/rpm/__db.xxx files - essentially this would force the file-backed shared memory regions to never be used.
- patch the version of librpm built/linked in osquery to never use the file-backed shared memory regions.

I'll include some repro steps in a follow-up comment.

**Comment by tomlanyon** [comment]
/bin/rpm works OK:
```
$ sudo rpm -qa | wc -l
580
```

osquery (non-root) can query rpm_packages, gets the same results, but throws a few warnings:
```
$ osqueryi --verbose "SELECT count(*) FROM rpm_packages ORDER BY name, arch, version, release;"
I0418 21:16:45.800508  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800572  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800595  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800618  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800640  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800662  1415 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 21:16:45.800686  1415 rpm_packages.cpp:53] Could not get RPM header flag.
+----------+
| count(*) |
+----------+
| 580      |
+----------+
```


same osquery invocation, but as root (notice the 'sudo'), can't open the RPM DB:
```
$ sudo osqueryi --verbose "SELECT count(*) FROM rpm_packages ORDER BY name, arch, version, release;"
error: db6 error(-30969) from dbenv->open: BDB0091 DB_VERSION_MISMATCH: Database environment version mismatch
error: cannot open Packages index using db6 -  (-30969)
error: cannot open Packages database in
error: db6 error(-30969) from dbenv->open: BDB0091 DB_VERSION_MISMATCH: Database environment version mismatch
error: cannot open Packages index using db6 -  (-30969)
error: cannot open Packages database in
+----------+
| count(*) |
+----------+
| 0        |
+----------+
```

rebuilding the RPM DB cleans up & removes the /var/lib/rpm/__db.xxx shared memory region files, and allows osquery (as root) to open RPM DB successfully:
```
$ sudo rpm --rebuilddb
$ sudo osqueryi --verbose "SELECT count(*) FROM rpm_packages ORDER BY name, arch, version, release;"
I0418 20:16:55.638063 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638109 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638126 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638140 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638155 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638176 29406 rpm_packages.cpp:53] Could not get RPM header flag.
I0418 20:16:55.638192 29406 rpm_packages.cpp:53] Could not get RPM header flag.
+----------+
| count(*) |
+----------+
| 580      |
+----------+
```

but this has written /var/lib/rpm/__db.xxx with the new db6 format and stops /bin/rpm being able to open the DB:
```
$ sudo rpm -qa | wc -l
error: db5 error(12) from dbenv->open: Cannot allocate memory
error: cannot open Packages index using db5 - Cannot allocate memory (12)
error: cannot open Packages database in /var/lib/rpm
error: db5 error(12) from dbenv->open: Cannot allocate memory
error: cannot open Packages database in /var/lib/rpm
0
```

until a RPM DB rebuild is performed again:
```
$ sudo rpm --rebuilddb
$ sudo rpm -qa | wc -l
580
```

**Comment by theopolis** [comment]
Thanks for the analysis @tomlanyon and sorry for allowing this to happen!

Your advice about dropping privileges is exactly how we avoided this in the past: https://github.com/facebook/osquery/blob/master/osquery/tables/system/linux/rpm_packages.cpp#L143

And we had a unit-test for this: https://github.com/facebook/osquery/blob/master/osquery/core/tests/posix/permissions_tests.cpp#L241

Which made sure we can drop and that when we drop we do not interrupt internal threads managed by dependencies. Unfortunately the test is only valid when run as `uid==0` and our CI in Jenkins does not run the tests as `root`. Bad news time, this test is currently failing, so we've recently added a regression that exists in the 2.4.0 release rendering the permissions dropping useless.

After a few minutes of binary search building (not as easy since I'm trying to replicate the dependency versions used at the time to eliminate confounds):
Works: 6ab460f5db6fa8fc64f1c6170bed746093daaebd
Fails: f5bcc66ee39af1cdd1a9a55455e8e1543ae3f13e

I'll bi-sect further, and we'll fix this bug ASAP! We should also investigate the RPM libs to see if we can do some API setup that doesn't generate the shared cache.

**Comment by tomlanyon** [comment]
Thanks for the quick investigation, @theopolis.

Ah, that explains why it works on the EL6 test boxes and not on EL7 - the EL6 boxes have osquery 2.3.2 and EL7 are running 2.3.4, so I guess the problem occurred somewhere between those two versions.

From what I saw looking through librpm, there was no way to force use of the private memory region when opening an RPM DB.

**Comment by theopolis** [comment]
That's a bummer, but this should put us back into a safe place: #3187 I'll hook up the CI tests so this never regresses again.
